### PR TITLE
[OPA-1889] Fix some of the remaining grok discrepancies

### PR DESCRIPTION
**Pull request by vladimir-dd**
Fixes some of the found grok discrepancies/bugs, comparing to Datadog [grok parser](https://docs.datadoghq.com/service_management/events/pipelines_and_processors/grok_parser/):

- `date` matcher - assume the current date, year when the time format does not include them.
- `keyvalue` filter - treat extracted fields as paths: `db.operation` means `{ "db": { "operation": "insert" }}`
-   drop `Value:Null` values from the parsed object

### CODE REVIEW

File: src/datadog/grok/filters/keyvalue.rs [hunk 1]
```diff
@@ -207,7 +207,7 @@ fn parse_line<'a>(
 }
 
 /// Parses the field_delimiter between the key/value pairs, ignoring surrounding spaces
-fn parse_field_delimiter<'a>(field_delimiter: &'a str) -> impl Fn(&'a str) -> SResult<&'a str> {
+fn parse_field_delimiter(field_delimiter: &str) -> impl Fn(&str) -> SResult<&str> + '_ {
```
Comment (pront) [inline src/datadog/grok/filters/keyvalue.rs:210]: nice, much easier to read now

File: src/datadog/grok/matchers/date.rs [hunk 2]
```diff
@@ -241,76 +243,122 @@ pub fn time_format_to_regex(format: &str, with_captures: bool) -> Result<RegexRe
...
+
+    if day_is_missing {
+        strp_format = format!("%-d {}", strp_format);
+    }
+    if month_is_missing {
+        strp_format = format!("%-m {}", strp_format);
+    }
+    if year_is_missing {
+        strp_format = format!("%Y {}", strp_format);
+    }
```
Comment (pront) [inline src/datadog/grok/matchers/date.rs:262]: I don't see any bugs here, however it is a complex piece of code. We can improve readability/maintainability by extracting this in a separate function e.g. 

```
pub fn adjust_strp_format(
    original_format: &str, 
    filter_format: &str
) -> String {
    let mut adjusted_format = filter_format.to_string();

    let year_is_missing = !original_format.contains('y') && !original_format.contains('Y');
    let month_is_missing = !original_format.contains('M');
    let day_is_missing = !filter_format.contains('d');

    if day_is_missing {
        adjusted_format = format!("%-d {}", adjusted_format);
    }
    if month_is_missing {
        adjusted_format = format!("%-m {}", adjusted_format);
    }
    if year_is_missing {
        adjusted_format = format!("%Y {}", adjusted_format);
    }

    adjusted_format
}
```

And then we can add a few tests e.g.
```
#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_adjust_strp_format() {
        assert_eq!(adjust_strp_format("", strp_format), "%-d %-m %Y ");
        // ...
    }
}
```

File: src/datadog/grok/matchers/date.rs [hunk 3]
```diff
@@ -241,76 +243,122 @@ pub fn time_format_to_regex(format: &str, with_captures: bool) -> Result<RegexRe
...
+
     let timestamp = match value {
         Value::Bytes(bytes) => {
-            let mut value = String::from_utf8_lossy(bytes).into_owned();
+            let original_value = String::from_utf8_lossy(bytes).into_owned();
```
Comment (pront) [inline src/datadog/grok/matchers/date.rs:266]: Same as above.

File: src/path/mod.rs [hunk 4]
```diff
@@ -261,6 +261,14 @@ impl<'a> ValuePath<'a> for &'a str {
     }
 }
 
+impl<'a> ValuePath<'a> for &'a OwnedValuePath {
```
Comment (pront) [inline src/path/mod.rs:264]: What was the motivation for moving this here?

Comment (vladimir-dd) [reply src/path/mod.rs:264]: I had an issue with the visibility of this trait implementation, but actually it should be in the PR for `keyvalue` filter - I'll double-check if I do need this and, if so, move to that PR. Reverting this change here.

File: src/datadog/grok/matchers/date.rs (hunk 2 above)
Comment (vladimir-dd) [reply src/datadog/grok/matchers/date.rs:262]: Good point, updated.

File: src/datadog/grok/matchers/date.rs [hunk 5]
```diff
@@ -241,120 +243,153 @@ pub fn time_format_to_regex(format: &str, with_captures: bool) -> Result<RegexRe
...
-                value.push_str("+0000");
-            }
-            if let Some(tz) = filter
-                .regex
-                .captures(&value)
-                .and_then(|caps| caps.name("tz"))
-            {
-                let tz = tz.as_str();
-                let tz: Tz = tz.parse().map_err(|error| {
+    let original_value = String::from_utf8_lossy(value.as_bytes().ok_or_else(|| {
```
Comment (vladimir-dd) [inline src/datadog/grok/matchers/date.rs:246]: The only real changes here is `adjust_strp_format_and_value()` function below.

File: src/datadog/grok/matchers/date.rs [hunk 6]
```diff
@@ -241,120 +243,141 @@ pub fn time_format_to_regex(format: &str, with_captures: bool) -> Result<RegexRe
...
+            .ok_or_else(|| {
+                GrokRuntimeError::FailedToApplyFilter(
+                    filter.to_string(),
+                    original_value.to_string(),
+                )
+            })?;
+        Ok(Value::from(
+            Utc.from_utc_datetime(&dt.naive_utc()).timestamp_millis(),
+        ))
+    } else {
```
Comment (neuronull) [inline src/datadog/grok/matchers/date.rs:290]: This is a totally non-blocking comment but an observation- there are a ton of nested if/else, if let  branches in this function and that makes it pretty hard to follow the logic when it gets so large. It'd be helpful to have either helper functions or a comment somewhere that describes the behavior.

File: src/datadog/grok/parse_grok.rs [hunk 7]
```diff
@@ -93,16 +95,51 @@ fn apply_grok_rule(source: &str, grok_rule: &GrokRule) -> Result<Value, Error> {
...
+        _ => {}
+    }
+}
+
 #[cfg(test)]
 mod tests {
     use crate::btreemap;
     use crate::value::Value;
+    use chrono::{DateTime, Datelike, Timelike};
+    // use chrono::{DateTime};
```
Comment (neuronull) [inline src/datadog/grok/parse_grok.rs:142]:
```suggestion
```

**Review by neuronull — APPROVED** [review]
Overall LGTM though I noticed CI failures. I appreciate the additional unit tests

File: src/datadog/grok/matchers/date.rs (hunk 6 above)
Comment (vladimir-dd) [reply src/datadog/grok/matchers/date.rs:290]: Totally agree - I added a [task](https://datadoghq.atlassian.net/browse/OPA-1935) to address this issue.
